### PR TITLE
[#13061][bugfix] Add rotate_activation to DSA fused_cat_fp8 path

### LINKED ISSUE #13061: [Bug]: fused_cat_fp8 didn't do rotate_activation for DSA
### System Info

System Info
TensorRT-LLM version: v1.3.0rc11
Model: deepseek v3.2

### Who can help?

_No response_

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [x] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

python examples/llm-api/quickstart_advanced.py \
  --model_dir DeepSeek-V3.2/ \
  --attention_backend TRTLLM \
  --moe_backend WIDEEP \
  --kv_cache_fraction 0.15 \
  --disable_overlap_scheduler \
  --disable_kv_cache_reuse \
  --tokens_per_block 64 \
  --max_seq_len 5000 \
  --max_num_tokens 4096 \
  --max_tokens 200 \
  --enable_chunked_prefill \
  --tp 8 \
  --prompt "$(cat long_context.txt)"

long_context.txt contains a single prompt about 4k tokens

### Expected behavior

Do rotate_activation after cat [qk_pe, qk_nope], then do quant

### actual behavior

Do cat then quant, but no rotate_activation

### additional notes

Although deepseek v3.2 is very robust, I think rotate_activation is still needed. @kaiyux 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by waynehacking8**
## Summary

Fixes #13061 — the DSA sparse attention indexer's `_prep_q_or_k` method calls `fused_cat_fp8` to concatenate `[qk_pe, qk_nope]` and quantize to FP8 in a single fused kernel, but **skips `rotate_activation`** (Hadamard rotation) which DSA requires between concatenation and quantization.

- **Root cause**: `fused_cat_fp8` is a single CUDA kernel that does cat+quant atomically — there's no way to inject the rotation in between.
- **Fix**: Unfuse the FP8 path into `torch.cat → rotate_activation → fp8_quantize_1x128`. The scale transpose logic follows the reference implementation in `test_fused_cat_fp8.py`.
- **Backward-compatible**: `rotate_activation` gracefully returns identity when `fast-hadamard-transform` is not installed (existing warning).
- **FP4 gap**: The FP4 path (`fused_cat_fp4`) has the same issue but needs a standalone FP4 quantize-only kernel to unfuse — noted in the docstring as a follow-up.

## Code flow (before → after)

**Before** (buggy):
```
qk_pe, qk_nope → fused_cat_fp8 (cat + quant) → fp8_out, scale
                   rotate_activation skipped
```

**After** (fixed):
```
qk_pe, qk_nope → torch.cat → rotate_activation → fp8_quantize_1x128 → fp8_out, scale
                               Hadamard rotation applied
```

## Test plan

- [x] `rotate_activation` fallback: returns identity when fast-hadamard-transform is missing — no regression for existing setups

Not run in my environment (no GPU build / no DeepSeek V3.2 weights) — flagged for CI / reviewer validation:

- `test_fused_cat_fp8.py`: the `fused_cat_fp8` CUDA kernel is **unchanged** by this PR (the FP8 path now reaches the quantizer via `fp8_quantize_1x128_sf_transpose` after the Hadamard rotation), so the kernel test is unaffected.
- DSA sparse attention end-to-end with DeepSeek V3.2 (requires model weights + `fast-hadamard-transform`).

Closes #13061
## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Summary by CodeRabbit

* **Refactor**
  * Optimized sparse attention backend with improved quantization implementation and enhanced tensor processing operations.

### CODE REVIEW

**Comment by waynehacking8** [comment]
Status check: this PR has been mergeable and review-ready since 06-09 (coderabbit pass addressed at open time). Could a maintainer trigger CI (`/bot run`) when convenient? Happy to rebase or adjust if anything in the recent refactors moved — the fix still applies cleanly to current main.

**Comment by karljang** [comment]
@waynehacking8,
Thank you for your contributions! Before we proceed, could you please address the pre-commit check failures?

**Comment by waynehacking8** [comment]
Thanks @karljang — addressed in e570480: the `_prep_q_or_k` import and `fp8_quantize_1x128_sf_transpose` call now match the repo's isort/yapf style (backslash-continued single import; closing paren on the last-argument line), which was the pre-commit failure. I also resolved the two open items in the PR checklist (the kernel-test one is unaffected since `fused_cat_fp8` is unchanged; the DeepSeek V3.2 end-to-end run needs model weights + a GPU I don't have locally, so I've flagged it for CI/reviewer validation). Ready for `/bot run` whenever you'd like.

**Comment by PerkzZheng** [comment]
Add @lfr-0531 to review as well. thanks

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py [hunk 1]
```diff
@@ -2326,18 +2326,23 @@ def _qk_projection_and_rope(self, qr: torch.Tensor, indexer_k: torch.Tensor,
...
         """
         if self.use_fp4:
             return torch.ops.trtllm.fused_cat_fp4(qk_pe, qk_nope)
-        fp8_out, scale = torch.ops.trtllm.fused_cat_fp8(
-            qk_pe, qk_nope, self.scale_fmt == "ue8m0")
-        return fp8_out, scale
+        from tensorrt_llm.quantization.utils.fp8_utils import \
+            fp8_quantize_1x128_sf_transpose
+        combined = torch.cat([qk_pe, qk_nope], dim=-1)
+        combined = rotate_activation(combined)
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/sparse/dsa.py]: This PR basically revert https://github.com/NVIDIA/TensorRT-LLM/pull/11899, which removes Hadamard rotation.
@kaiyux as the author, could you tell us why we can remove Hadamard rotation here?

**Comment by waynehacking8** [comment]
@yuxianq @PerkzZheng Thanks for the careful review. I looked into the history — this isn't an intentional revert of a design choice, it's restoring a regression:

#11899 was `[perf]`-only: it fused cat+quantize into `fusedCatFp8.cu`. That kernel does **only** concat (Stage 1) + FP8 quantize (Stage 2) — no Hadamard rotation — and its `dsa.py` diff deleted `q_or_k = rotate_activation(q_or_k)` without re-adding it anywhere. So the rotation was dropped as a side effect of the fusion, not by design; #13061 is exactly that silent regression.

This PR restores the pre-#11899 fp8 path. I've aligned the fp8 branch to mirror the original implementation verbatim — `maybe_compiled_cat` → `rotate_activation` → `fp8_utils.fp8_quantize_1x128_sf_transpose` — so it's byte-for-byte the code #11899 replaced (only `fused_cat_fp4` is kept for the fp4 case). Same Hadamard, same quantizer ⇒ numerically the original intended behavior. Rotate-before-FP8 matters for quant accuracy (Hadamard smooths outliers); DSV3.2 is robust enough to mask it, but it's still a correctness drop.

The cost is that un-fusing gives up #11899's kernel speedup. Happy to follow up by folding the rotation into `fusedCatFp8.cu` (and the FP4 path, which still lacks it) to recover perf — separate PR or in this one, whichever you and @kaiyux prefer.

File: tensorrt_llm/_torch/attention_backend/sparse/dsa.py (hunk 1 above)
Comment (waynehacking8) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa.py]: Replied in detail in a top-level comment — TL;DR: this isn't a revert. #11899 was a `[perf]` fusion, and `fusedCatFp8.cu` only does concat + FP8 quantize (no Hadamard), so the rotation was dropped unintentionally (that's #13061). I've restored the pre-#11899 `cat → rotate → quantize` path verbatim (same `rotate_activation`, same `fp8_utils.fp8_quantize_1x128_sf_transpose`). Happy to fold the rotation into the kernel as a perf follow-up.

Comment (yuxianq) [reply tensorrt_llm/_torch/attention_backend/sparse/dsa.py]: I find a comment about it: https://github.com/NVIDIA/TensorRT-LLM/pull/11899#discussion_r2903683110, which says:

> Yes, I intentionally removed it. `hadamard_transform` was not used, and previous verification shows that it's safe to remove with no accuracy drop.

@waynehacking8 Can you show us a case that accuracy drop for a benchmark without Hadamard rotation?

**Comment by waynehacking8** [comment]
@yuxianq Thanks — kaiyux's #11899 comment is the decisive context. I worked through it and his call holds.

The indexer logit is `out_ij = Σ_h ReLU(q[i,h]·k[j]) · weights[i,h]`, and `_prep_q_or_k` applies the same orthonormal Hadamard H to both q and k (dsa.py:2376-2377). Since `(Hq)·(Hk) = q·k`, the per-head dot product — and therefore the ReLU, the weights, and the summed logit — are all unchanged in exact arithmetic. The Hadamard's only real effect is lowering FP8 quantization error on that dot product before it's computed (the QuaRot rationale), and kaiyux's verification shows that doesn't move the indexer's top-k selection.

I can't produce a config with an actual accuracy drop (no GPU / DSV3.2 weights on my side), and given the cancellation I don't expect one for the FP8 path. So I'll defer to the verification and close this — #13061 is really a question, not a regression. If FP4's larger quant error ever turns out to move top-k, that'd be the place to revisit.

Thanks all for the quick reviews.
